File: p036.py
from itertools import count, takewhile
LIMIT = 10**6


def decimal_palindromes():
    for i in count(1):
        for n in range(10**(i-1), 10**i):
            yield int(str(n) + str(n)[-2::-1])
        for n in range(10**(i-1), 10**i):
            yield int(str(n) + str(n)[::-1])


def bin_is_pali(n):
    binstr = bin(n)[2:]
    return binstr == binstr[::-1]


ans = sum(filter(bin_is_pali, takewhile(lambda n: n < LIMIT,
                                        decimal_palindromes())))

print(ans)

File: p036.scala
val LIMIT = math.pow(10, 6).toInt

def merge (a: Stream [Int], b: Stream [Int]): Stream [Int] = 
  if (a.head<b.head) a.head #:: merge(a.tail, b)
  else b.head #:: merge(b.tail, a)

def decimalPalindromes(): Stream[Int] = merge (
  Stream from 1 map (n =>
    Integer.parseInt((n/10).toString + n.toString.reverse)),
  Stream from 1 map (n =>
    Integer.parseInt(n.toString + n.toString.reverse)))

def binIsPali(n: Int) = {
  val binstr = Integer.toBinaryString(n)
  binstr equals binstr.reverse
}

val ans = decimalPalindromes.takeWhile(_ < LIMIT).filter(binIsPali(_)).sum

println(ans)
